File: problems/509_fibonacci-number/main.cpp
#include <iostream>
#include <utility>
#include <vector>

class Solution
{
public:
    std::vector<int> dp;
    int fibrec(int n)
    {
        if (dp[n] != -1)
            return dp[n];
        if (n == 0 or n == 1)
            dp[n] = n;
        else
            dp[n] = fibrec(n - 1) + fibrec(n - 2);
        return dp[n];
    }
    int fib_3(int n)
    {
        if (n == 0 || n == 1)
            return n;
        int prev_1 = 0, prev_2 = 1, fibo = 1;
        for (int i = 2; i <= n; i++)
        {
            fibo = prev_1 + prev_2;
            prev_1 = prev_2;
            prev_2 = fibo;
        }
        return fibo;
    }
    int fib_2(int n)
    {
        dp.assign(n + 1, -1);
        return fibrec(n);
    }
    int fib(int n)
    {
        if (n == 0 || n == 1)
            return n;
        return fib(n - 1) + fib(n - 2);
    }
};

int main()
{
    Solution solution;
    std::vector<std::pair<int, int>> samples = {
        {2, 1},
        {3, 2},
        {4, 3},
    };

    for (const auto &[n, expected] : samples)
    {
        int result = solution.fib(n);
        std::cout << "input: n = " << n
                  << ", placeholder result: " << result
                  << ", expected: " << expected << '\n';
    }

    return 0;
}

File: problems/509_fibonacci-number/main.rs
struct Solution;

impl Solution {
    pub fn fib(n: i32) -> i32 {
        if n == 0 || n == 1 {
            return n;
        }
        Self::fib(n - 1) + Self::fib(n - 2)
    }

    pub fn fib_2(n: i32) -> i32 {
        let mut dp = vec![-1; (n + 1) as usize];
        Self::fibrec(n, &mut dp)
    }

    fn fibrec(n: i32, dp: &mut Vec<i32>) -> i32 {
        if dp[n as usize] != -1 {
            return dp[n as usize];
        }
        if n == 0 || n == 1 {
            dp[n as usize] = n;
        } else {
            dp[n as usize] = Self::fibrec(n - 1, dp) + Self::fibrec(n - 2, dp);
        }
        dp[n as usize]
    }

    pub fn fib_3(n: i32) -> i32 {
        if n == 0 || n == 1 {
            return n;
        }
        let mut prev_1 = 0;
        let mut prev_2 = 1;
        let mut fibo = 1;
        for _ in 2..=n {
            fibo = prev_1 + prev_2;
            prev_1 = prev_2;
            prev_2 = fibo;
        }
        fibo
    }
}

fn main() {
    let samples = vec![(2, 1), (3, 2), (4, 3)];

    for (n, expected) in samples {
        let result = Solution::fib(n);
        println!(
            "input: n = {}, placeholder result: {}, expected: {}",
            n, result, expected
        );
    }
}
